File: OrderManagementService/Models/Order.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System;
using System.Collections.Generic;

namespace OrderManagementService.Models
{
    
    public class Order
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; }

        [BsonElement("userId")]
        public string UserId { get; set; }

        [BsonElement("orderNumber")]
        public string OrderNumber { get; set; }

        [BsonElement("shippingAddress")]
        public ShippingAddress ShippingAddress { get; set; }

        [BsonElement("items")]
        public List<OrderItem> Items { get; set; }

        [BsonElement("totalAmount")]
        public decimal TotalAmount { get; set; }

        [BsonElement("discountAmount")]
        public decimal DiscountAmount { get; set; } = 0;

        [BsonElement("couponCode")]
        public string CouponCode { get; set; }

        [BsonElement("loyaltyPointsUsed")]
        public int LoyaltyPointsUsed { get; set; } = 0;

        [BsonElement("loyaltyPointsEarned")]
        public int LoyaltyPointsEarned { get; set; } = 0;

        [BsonElement("status")]
        public string Status { get; set; } = "pending";

        [BsonElement("statusHistory")]
        public List<OrderStatusHistory> StatusHistory { get; set; }

        [BsonElement("createdAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [BsonElement("updatedAt")]
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    }

    public class ShippingAddress
    {
        [BsonElement("receiverName")]
        public string ReceiverName { get; set; }

        [BsonElement("phoneNumber")]
        public string PhoneNumber { get; set; }

        [BsonElement("addressLine")]
        public string AddressLine { get; set; }

        [BsonElement("ward")]
        public string Ward { get; set; }

        [BsonElement("district")]
        public string District { get; set; }

        [BsonElement("city")]
        public string City { get; set; }
    }


}

File: frontend/lib/models/Order.dart
import 'OrderItem.dart';
import 'OrderStatusHistory.dart';
import 'ShippingAddress.dart';

class Order {
  final String? id;
  final String? userId;
  final String orderNumber;
  final ShippingAddress shippingAddress;
  final List<OrderItem> items;
  final double totalAmount;
  final double discountAmount;
  final String? couponCode; // Nullable, giữ giống C#
  final int loyaltyPointsUsed;
  final int loyaltyPointsEarned;
  final int shippingFee;
  String status;
  List<OrderStatusHistory> statusHistory;
  final DateTime createdAt;
  DateTime updatedAt;

  Order({
    this.id,
    this.userId,
    required this.orderNumber,
    required this.shippingAddress,
    required this.items,
    required this.totalAmount,
    this.discountAmount = 0,
    this.couponCode, // Có thể null
    this.loyaltyPointsUsed = 0,
    this.loyaltyPointsEarned = 0,
    this.shippingFee = 30000,
    this.status = 'pending',
    required this.statusHistory,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) => Order(
    id: json['_id'] ?? json['id'], // Có thể là _id hoặc id
    userId: json['userId'],
    orderNumber: json['orderNumber'],
    shippingAddress: ShippingAddress.fromJson(json['shippingAddress']),
    items: (json['items'] as List).map((e) => OrderItem.fromJson(e)).toList(),
    totalAmount: (json['totalAmount'] as num).toDouble(),
    discountAmount: (json['discountAmount'] as num?)?.toDouble() ?? 0.0,
    couponCode: json.containsKey('couponCode') ? json['couponCode'] : null,
    loyaltyPointsUsed: json['loyaltyPointsUsed'] ?? 0,
    loyaltyPointsEarned: json['loyaltyPointsEarned'] ?? 0,
    shippingFee: json['shippingFee'] ?? 30000,
    status: json['status'] ?? 'pending',
    statusHistory: (json['statusHistory'] as List)
        .map((e) => OrderStatusHistory.fromJson(e)).toList(),
    createdAt: DateTime.parse(json['createdAt']),
    updatedAt: DateTime.parse(json['updatedAt']),
  );

  Map<String, dynamic> toJson() => {
    '_id': id,
    'userId': userId,
    'orderNumber': orderNumber,
    'shippingAddress': shippingAddress.toJson(),
    'items': items.map((e) => e.toJson()).toList(),
    'totalAmount': totalAmount,
    'discountAmount': discountAmount,
    // Luôn gửi couponCode, có thể là '' hoặc null nếu không có
    'couponCode': couponCode ?? '',
    'loyaltyPointsUsed': loyaltyPointsUsed,
    'loyaltyPointsEarned': loyaltyPointsEarned,
    'shippingFee': shippingFee,
    'status': status,
    'statusHistory': statusHistory.map((e) => e.toJson()).toList(),
    'createdAt': createdAt.toIso8601String(),
    'updatedAt': updatedAt.toIso8601String(),
  };
}
